### ISSUE TITLE
Design annotation integration with Muse markup

### ISSUE THREAD

**Issue by bibliotechie** [bibliotechie/antithesis#3]
This will likely be informed by the [target API](https://h.readthedocs.io/en/latest/api-reference/v1/) (as well as the existing[ Muse markup syntax](https://amusewiki.org/library/manual)).

**Comment by bibliotechie** [comment]
The original goal was to integrate the annotations directly into the Muse markup. There are two main reasons for this:
 - It allows using the Muse markup to compile the annotations statically into LaTeX (and therefore other formats like PDF and epub).
 - It allows us to track conflicts between changes from the upstream Muse files and the local annotations using git and the available associated tools.

However, there are some major drawbacks to this approach as well:
 - All annotations have to be fully moderated, just like the texts are.
 - Annotations are represented in the API with a moderately complicated JSON object (15 child objects, many of which expand into other nested objects, up to 3 layers deep). Muse supports XML-style tags, but translating whole JSON objects like that into an XML tree seems a lot heavier than the syntax Muse markup was intended to handle (keep in mind, we'll probably end up with dozens of annotations on larger texts).
 - Lookup of annotations requires text parsing instead of DB lookups, making it slower and more complicated.

My current thought is therefore some form of compromise, where we store most of the annotation data in a DB, and keep only the information needed to statically render the annotations within the muse text (along with the ID to correlate them). This is just my current thinking though, I don't have a specific design in mind.

**Comment by bibliotechie** [comment]
We should possibly consult upstream on this issue once we have some more concrete ideas, see what they think.

**Comment by cohosh** [comment]
> We should possibly consult upstream on this issue once we have some more concrete ideas, see what they think.

By upstream, I'm assuming you mean amusewiki?

**Comment by cohosh** [comment]
Cool. I agree that a compromise where we store the full annotation in a DB and a subset of the data in markup is a good plan.

Just to be clear: we'd be making an extension to the Muse markup (i.e., new tags) to store the annotations? And then we have the future task of figuring out how to render the annotations with LaTeX.

Looking at how an annotation is represented from the [hypothesis API search response](https://github.com/bibliotechie/antithesis/wiki/Hypothesis-API%3A-search#what-it-looks-like), it looks like we'd really just want to represent the following in the markup:
- text
- group (depends on whether we want to display different groups differently or perhaps use this to decide whether or not to display the annotation)
- replies
- user_info->display_name
and then use the `target` information (along with values of `hidden` or `permissions`) to determine if it appears and where to place the markup in the document?

**Comment by mycelian** [comment]
Cool. I agree that a compromise where we store the full annotation in a DB and a subset of the data in markup is a good plan.

Just to be clear: we'd be making an extension to the Muse markup (i.e., new tags) to store the annotations? And then we have the future task of figuring out how to render the annotations with LaTeX.

Looking at how an annotation is represented from the [hypothesis API search response](https://github.com/bibliotechie/antithesis/wiki/Hypothesis-API%3A-search#what-it-looks-like), it looks like we'd really just want to represent the following in the markup:
- text
- group (depends on whether we want to display different groups differently or perhaps use this to decide whether or not to display the annotation)
- replies
- user_info->display_name

and then use the `target` information (along with values of `hidden` or `permissions`) to determine if it appears and where to place the markup in the document?

**Comment by bibliotechie** [comment]
> Just to be clear: we'd be making an extension to the Muse markup (i.e., new tags) to store the annotations? And then we have the future task of figuring out how to render the annotations with LaTeX.

Yep, that's the idea.
 
>     * group (depends on whether we want to display different groups differently or perhaps use this to decide whether or not to display the annotation)

I know we haven't pinned down the design exactly, but I think there would only be a need for one group for the annotations embedded into the text, so we could exclude this. I can't currently see a good use-case for having the group feature while also allowing annotations from different groups be embedded into the muse markup itself. I.e., I think if we were to implement groups, the annotations associated with those groups could be stored purely in the DB.

>     * replies

Depending on what design we land on, this could be represented by just nesting the replies in the annotation body.

> and then use the `target` information (along with values of `hidden` or `permissions`) to determine if it appears and where to place the markup in the document?

Right now, my theory is that we can find a way to encode that sort of thing implicitly. 

I also think we should include the annotation ID, and maybe one of "created" or "updated".

Here's [the Amuse markup for amusewiki](www.amusewiki.org/library/manual), and [the original Muse markup for emacs](https://www.gnu.org/software/emacs-muse/manual/muse.html). A major difference between the two I'll note is that while the original Muse has tags that have attributes (e.g., `<cite type="author">Miller1999</cite>`), Amuse does not.

I have a few ideas for how to structure this, but they still need to be fleshed out more. The theme  of the ones I like the most though are treating annotations as a third form of [footnote](https://amusewiki.org/library/manual#toc16) (the other two extant being the original square bracket and secondary curly brace, see the documentation in the link). The goal would be for the main difference between an annotation and the other footnote types to be that annotations are associated with a text range, rather than a singular point within the text. The body of the annotation would be associated to the inline reference via the annotation ID. The obvious way of representing the inline reference would be something like a `<annotate id=foobar>Some text...</annotate>`, then later in the text `<annotation id=foobar author=Anonymous created=2020-01-01>Insightful comment</annotation>`. The problem with this design being it relies heavily on attributes, which, again, would be a new thing for Amuse syntax. If we don't want to add attributes, the other idea I had was using `<annotate>...</annotate>@idval@` and `@idval@ Insightful comment`, i.e., using `@` (or whatever other uncommon character not already used in the Amuse syntax) as a new footnote delimiter, one that automatically associates with the last `<annotation>` tag. The problem with this is it's a little weird, and is obviously a hack to get around attributes.

Anyways, that's where my thought process is so far. We should discuss this more.

**Comment by bibliotechie** [comment]
Oh right, forgot to add: another problem with the delimiter-style is there's nowhere to put the author or creation date. We could work around that by just including that information in the body, as normal text in the annotation, formatted with normal Amuse markup, but it would be nice to separate that sort of information more cleanly.

**Comment by bibliotechie** [comment]
I just came up with another syntax that I think works better than either of the above. The idea is something analogous to the Amuse syntax for [floating images](https://amusewiki.org/library/manual#toc30), but with curly braces `{{}}` instead of square brackets `[[]]`. So it would look something like this:
```
First sentence. {{Second sentence}{}. Third sentence.

**Comment by bibliotechie** [comment]
I just came up with another two syntaxes, inspired by the Amuse syntax for [links](https://amusewiki.org/library/manual#toc28), but with curly braces `{{}}` instead of square brackets `[[]]`. The first option is one directly modeled after the link syntax:
```
First sentence. {{Insightful comment.}{Second sentence.}} Third sentence.
```
(This keeps the `[[target][source]]` order from links, which looks backwards; we'd need to decide whether we prefer consistency or readability).
This could be augmented with positional attributes like [floating images](https://amusewiki.org/library/manual#toc30), though we'd need a separator other than space:
```
First sentence. {{Insightful comment.|id|author|date}{Second sentence.}} Third sentence.
```
The second option is to basically mix the XML, delimiter, and link syntax:
```
First sentence. {{id}{Second sentence.}} Third sentence.

{{id|author|date}} Insightful comment
```
This would look closer to a footnote, and makes the annotation body its own text rather than an attribute.

The disadvantage of these being the lack of named attributes in favor of positional ones, meaning we'd have to be careful to keep our attributes in line with upstream attributes.

**Comment by bibliotechie** [comment]
I think I'm actually starting to lean towards the last one, but dropping the separators (so just `{{id}} Insightful comment.`) and just accept that representing that sort of metadata within the markup syntax is stretching it too thin. We can just emphasize it ourselves within the body.

**Comment by mycelian** [comment]
I like both of these better than the previous two proposals.

> I just came up with another two syntaxes, inspired by the Amuse syntax for [links](https://amusewiki.org/library/manual#toc28), but with curly braces `{{}}` instead of square brackets `[[]]`. The first option is one directly modeled after the link syntax:
> 
> ```
> First sentence. {{Insightful comment.}{Second sentence.}} Third sentence.
> ```
> 
> (This keeps the `[[target][source]]` order from links, which looks backwards; we'd need to decide whether we prefer consistency or readability).

FWIW I actually think
```
First sentence. {{Second sentence.}{Insightful comment.}} Third sentence.
```
is also consistent with the link syntax (and more readable ofc), since the comment can be thought of as a description of the target sentence. I see that it depends on how you look at it though.

**Comment by mycelian** [comment]
> I think I'm actually starting to lean towards the last one, but dropping the separators (so just `{{id}} Insightful comment.`) and just accept that representing that sort of metadata within the markup syntax is stretching it too thin. We can just emphasize it ourselves within the body.

I agree with this. I like the footnote style better because it makes reading the markup a bit easier.

I also like the idea of removing authorship and date from the markup itself. To carry the margin notes book analogy, you might not necessarily be able to tell who wrote what in a book when (although you would see differences in handwriting/fading).

**Comment by bibliotechie** [comment]
Cool, waiting for feedback from melmothx/text-amuse/issues/61 :)
